### PR TITLE
Add shopping cart to POS

### PR DESCRIPTION
**Pull request by mariodian**
Hi, I've created a fully client-side shopping cart for POS that stores the data in the browser's Local Storage.

When the cart is enabled, custom pricing is disabled because it would make the cart's content look messy.

screen shot 2018-11-20 at 3 26 29 pm

screen shot 2018-11-20 at 3 26 33 pm

P.S. I haven't used javascript for quite a long time so I'm not up to date with the latest conventions and whatnot. Please let me know if I should do things differently.
P.S2 I haven't figured out how to use language strings in the view instead of hardcoded English.
P.S3. I don't know how compatible with old browser we wanna be so I haven't used the latest javascript features such as proper Class declaration etc. Let me know if something needs to be done differently.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
@mariodian actually this would go well with this PR https://github.com/btcpayserver/btcpayserver/pull/409/files . It would basically show the item list properly in the Invoice view. :)

**Comment by NicolasDorier** [comment]
Does not pass CI, re-running.

**Comment by Kukks** [comment]
Language strings are unfortunately only on the checkout page for now :/
We don't use any form of transpilation so js has to be as ancient compat as possible.

**Comment by mariodian** [comment]
> Does not pass CI, re-running.

Seems like the issue isn't with my code:

```
[xUnit.net 00:01:59.44]     BTCPayServer.Tests.UnitTest1.CanGetRateCryptoCurrenciesByDefault [FAIL]
Failed   BTCPayServer.Tests.UnitTest1.CanGetRateCryptoCurrenciesByDefault
Error Message:
 Impossible to get the rate ERR_RATE_UNAVAILABLE(zaif, MONA_BTC) * 4592.40557902933
```

> Language strings are unfortunately only on the checkout page for now :/

Ok, good to know.

**Comment by NicolasDorier** [comment]
pure awesome, will test it! @rockstardev it might interest you as well.

We can add the item list as posdata later after https://github.com/btcpayserver/btcpayserver/pull/409 get merged.

**Comment by NicolasDorier** [comment]
@mariodian if not too complicated, small nit: I think that arrows instead of textbox for the number of items is better. People using this PoS will likely do it on tablet where keyboard is a pain.

**Comment by mariodian** [comment]
Yeah, I was thinking about it. But then the same issue is with custom prices which use the same input.

It's not difficult but it will require a custom logic.

**Comment by NicolasDorier** [comment]
mmh actually might be ok. If they want more, they can click again on the same button. If they want to cancel, they just have to open the cart.

**Comment by mariodian** [comment]
That's how I'd been testing it but the +- buttons definitely make sense. 

Since they'd have to be implemented outside of the cart too it should be done as a separate PR, though.

**Comment by mariodian** [comment]
Now I'm thinking that the whole POS should be redesigned at some later stage so the items in the cart are visible all the time for better UX. 

Like this:

**Comment by Kixunil** [comment]
Hey, thanks for your work, this will be very useful to us at Paralelna Polis!

> I don't know how compatible with old browser we wanna be so I haven't used the latest javascript

I'd be glad if it was compatible with the old browser in Android 4.0 as we have such terminal in PP. I painfully found out it doesn't even support `var x = new Map();`

> Now I'm thinking that the whole POS should be redesigned at some later stage so the items in the cart are visible all the time for better UX.

This would be helpful as current (very old and messy) PoS does the same (items on the left side)

I didn't have time to review the code yet, but heads up to double-check summing of prices. I've already seen bugs in production using floats... I'm not sure what's the best way to handle it in JS (since every number is float there) - in my PoC code I stringly multiplied EUR prices by 100 and then summed cents.

**Comment by ketominer** [comment]
+1 for doing fixed point maths!

File: BTCPayServer/wwwroot/cart/js/cart.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        self = this,
+        list = [];
+    
+    if (this.content.length > 0) {
+        // Prepare the list of items in the cart
+        for (var key in this.content) {
+            let item = this.content[key];
+            let tableRow = '<tr data-id="' + item.id + '">' + 
+                        (item.image !== null ? '<td class="align-middle pr-0" width="60"><img src="' + item.image + '" width="100%"></td>' : '') +
+                        '<td class="align-middle pr-0"><b>' + item.title + '</b></td>' +
```
Comment (NicolasDorier) [inline BTCPayServer/wwwroot/cart/js/cart.js]: `item.title` is weak to javascript injection

**Comment by NicolasDorier** [comment]
Can you fix the place where you concat html with js properties so JS can't be injected?

**Comment by rockstardev** [comment]
@mariodian @NicolasDorier just heads up - I've rebased pull request that resolves that MONA_BTC exchange rate causing CircleCI to fail.

I'll also reach out to CircleCI to see if they can put us free open source plan with more build minutes.

File: BTCPayServer/wwwroot/cart/js/cart.js (hunk 1 above)
Comment (mariodian) [reply BTCPayServer/wwwroot/cart/js/cart.js]: Any example how you'd inject anything in it?

Comment (mariodian) [reply BTCPayServer/wwwroot/cart/js/cart.js]: I can use something similar to esc() function in paybutton.js but I tried injecting js to that code via the admin and seems like the server escapes everything already.

**Comment by mariodian** [comment]
I'll fix the flaot prices too.

**Comment by Kukks** [comment]
It could be a tip amount that they want to leave? :)

**Comment by NicolasDorier** [comment]
Ah right tip....

**Comment by mariodian** [comment]
Was thinking about adding the tip field in the cart.

**Comment by mariodian** [comment]
Added a tip input in the cart https://www.youtube.com/watch?v=v3AyNX447jU

Let me guys know if you want me to push the change.

**Comment by mariodian** [comment]
> That's cool, is it possible to change the text on it through the pos?
Good idea, I'll do it tmrw.

> I wouldn't remove the custom amount amount entries in any case in case you have a "pay as much as you want" kind of item
Make sense

**Comment by NicolasDorier** [comment]
I thought no custom amount because the shopping cart will eventually be save inside posData. 

Also, how do you show same items which have different prices? This seem lot's of unnecessary for a very marginal use case.

**Comment by mariodian** [comment]
> How would you show same items which have different prices? in the cart, and in the posData? This seem lot's of considerations for a very marginal use case

Custom prices per item are disabled when the cart is enabled. Yes, it would be a mess.

> I thought no custom amount...

We already have a switch for it in the administration ("User can input custom amount") so perhaps we should let the owner of the store decide?

**Comment by NicolasDorier** [comment]
This seems a mess. How would you render user's custom amount inside the cart?

Or maybe what we can do is: If this box is checked and cart is available, do not show it in the page, but inside the cart as you did as "Tip".

Another solution: Just disabling it for now (no custom item, no custom amount, no tip) and see in another PR?

**Comment by mariodian** [comment]
> This seems a mess. How would you render user's custom amount inside the cart?

You wouldn't, you would just pay straight away as you do now. The functionality of it wouldn't change. 

If the admin of POS thinks it will be a mess they can always disable it in the settings. But I don't think we should try to be "too smart" for them. There may be a use case where they want both the cart and the custom amount enabled at the same time. 

When it comes to custom amounts **for items**, that indeed should be disabled because there's no meaningful way to display it in the cart.

**Comment by NicolasDorier** [comment]
ok, should I merge as is then?

**Comment by mariodian** [comment]
I’ll add the custom text for the tip and you can merge.

**Comment by mariodian** [comment]
@NicolasDorier pushed the last commits, please have a look at it. 

Summary:
- new POS cart
- custom prices for items are disabled
- custom amount stays user-configurable
- tipping is enabled
- custom tip text

**Comment by NicolasDorier** [comment]
Sorry for this, can you rebase one more time? I broke the test CanUsePos by mistake.

**Comment by mariodian** [comment]
Rebased

**Comment by NicolasDorier** [comment]
this animation is dangerously addictive.

**Comment by KanoczTomas** [comment]
@NicolasDorier @mariodian  I tried switching to the newest tag with this pull request, but get an error for the existing app. Not even the settings page works for it. I know I could switch back to old tag, export the app and grt to the new one. I think this should be a convenient thing dine in the backround. Cinverting the existing app to support shopping card.

The stack:
```
fail: Microsoft.AspNetCore.Server.Kestrel: Connection id "0HLIK85IBRO11", Request id "0HLIK85IBRO11:00000013": An unhandled exception was thrown by the application.
Microsoft.AspNetCore.Mvc.Razor.Compilation.CompilationFailedException: One or more compilation failures occurred:
50hcncb2.x3r(964,1457): error CS1031: Type expected
50hcncb2.x3r(964,1691): error CS1031: Type expected
50hcncb2.x3r(964,3773): error CS1031: Type expected
50hcncb2.x3r(964,3997): error CS1031: Type expected
50hcncb2.x3r(964,4221): error CS1031: Type expected
50hcncb2.x3r(964,1518): error CS1061: 'UpdatePointOfSaleViewModel' does not contain a definition for 'EnableShoppingCart' and no extension method 'EnableShoppingCart' accepting a fi
st argument of type 'UpdatePointOfSaleViewModel' could be found (are you missing a using directive or an assembly reference?)
50hcncb2.x3r(964,1752): error CS1061: 'UpdatePointOfSaleViewModel' does not contain a definition for 'EnableShoppingCart' and no extension method 'EnableShoppingCart' accepting a fi
st argument of type 'UpdatePointOfSaleViewModel' could be found (are you missing a using directive or an assembly reference?)
50hcncb2.x3r(964,3829): error CS1061: 'UpdatePointOfSaleViewModel' does not contain a definition for 'CustomTipText' and no extension method 'CustomTipText' accepting a first argume
t of type 'UpdatePointOfSaleViewModel' could be found (are you missing a using directive or an assembly reference?)
50hcncb2.x3r(964,4053): error CS1061: 'UpdatePointOfSaleViewModel' does not contain a definition for 'CustomTipText' and no extension method 'CustomTipText' accepting a first argume
t of type 'UpdatePointOfSaleViewModel' could be found (are you missing a using directive or an assembly reference?)
50hcncb2.x3r(964,4277): error CS1061: 'UpdatePointOfSaleViewModel' does not contain a definition for 'CustomTipText' and no extension method 'CustomTipText' accepting a first argume
t of type 'UpdatePointOfSaleViewModel' could be found (are you missing a using directive or an assembly reference?)
   at Microsoft.AspNetCore.Mvc.Razor.Internal.RazorViewCompiler.CompileAndEmit(RazorCodeDocument codeDocument, String generatedCode)
   at Microsoft.AspNetCore.Mvc.Razor.Internal.RazorViewCompiler.CompileAndEmit(String relativePath)
   at Microsoft.AspNetCore.Mvc.Razor.Internal.RazorViewCompiler.OnCacheMiss(String normalizedPath)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Razor.Internal.DefaultRazorPageFactoryProvider.CreateFactory(String relativePath)
   at Microsoft.AspNetCore.Mvc.Razor.RazorViewEngine.CreateCacheResult(HashSet`1 expirationTokens, String relativePath, Boolean isMainPage)
   at Microsoft.AspNetCore.Mvc.Razor.RazorViewEngine.OnCacheMiss(ViewLocationExpanderContext expanderContext, ViewLocationCacheKey cacheKey)
   at Microsoft.AspNetCore.Mvc.Razor.RazorViewEngine.LocatePageFromViewLocations(ActionContext actionContext, String pageName, Boolean isMainPage)
   at Microsoft.AspNetCore.Mvc.Razor.RazorViewEngine.FindView(ActionContext context, String viewName, Boolean isMainPage)
   at Microsoft.AspNetCore.Mvc.ViewEngines.CompositeViewEngine.FindView(ActionContext context, String viewName, Boolean isMainPage)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.FindView(ActionContext actionContext, ViewResult viewResult)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)
   at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultAsync(IActionResult result)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResultFilterAsync[TFilter,TFilterAsync]()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResultExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultFilters()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at BTCPayServer.Hosting.BTCPayMiddleware.Invoke(HttpContext httpContext) in /home/ispol/bin/btcpayserver/BTCPayServer/Hosting/BTCpayMiddleware.cs:line 46
   at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

**Comment by mariodian** [comment]
> if you want create a donation page for you to showcase when I announce it on twitter

@NicolasDorier nah, it's alright :)

**Comment by NicolasDorier** [comment]
are you using the latest docker image @KanoczTomas ? how can I replicate this error?

**Comment by KanoczTomas** [comment]
@NicolasDorier I do not use docker, using manual setup on Ubuntu 18.04.1 LTS with dotnet-sdk-2.1.

I tried replicating it on a new `APP`, similar errors. Here is what I did (having the lattest pull):
  1. git checkout v1.0.3.17
  2. go to web admin interface and create a new APP (just leave the template as is).
  3. stop the server 
  4. git checkout v1.0.3.18
  5. navigate to POS app. It will not load the page and the logs will show the above mentioned error

Here is the btcpayserver config(obfuscated):
```
network=testnet
port=23001
bind=0.0.0.0
chains=btc,ltc
BTC.explorer.url=http://127.0.0.1:24445
LTC.explorer.url=http://127.0.0.1:24445
postgres=User ID=verysecretuser;Password=verysecretpassword;Host=localhost;Port=5432;Database=btcpayserver;
externalurl=http://verystealthyexternalurl.com:23001
```

**Comment by mariodian** [comment]
Yeah when you look at commits under the latest tag the code isn’t part of it.

**Comment by KanoczTomas** [comment]
@mariodian, nah it works with the latest tag as well. I am just so used to JS nowadays, that I forgot to run ./build.sh :) sorry

I was playing with it and realized when I have bellow options enabled for an APP the popup to pay does not appear once I click on `pay` in the cart. `btcpayserver` does not mention any new invoice creation, fail or any other message that would indicate it tried to create the invoice. When I enable `user can input custom amount` the payment screen works and show a QR code + invoice generation can be seen in `btcpayserver` logs.

image

I did a `git clean -xdf` then `git checkout 183169` (your commit) then `./build.sh` and `./run.sh`

**Comment by darkban** [comment]
how to install in my server ubuntu this funtion? i update my server in version ***.17 but not show this funtion

**Comment by NicolasDorier** [comment]
@darkban I still have not pushed it on the docker deploy.

**Comment by NicolasDorier** [comment]
@KanoczTomas can you check if you can reproduce on `http://mainnet.demo.btcpayserver.org`

**Comment by KanoczTomas** [comment]
@NicolasDorier https://mainnet.demo.btcpayserver.org/apps/RhAuWhcx3QmdNmMsfMR5fSJZmVr/pos

**Comment by mariodian** [comment]
I get the same behavior on my store. Except that mine doesn't even redirect me back but stays on the blank page.

http://btcpay.freedomnode.com:23001/apps/JpjYE5UVtteBVQRPY1MgzNvm5qY/pos

**Comment by mariodian** [comment]
I can't even create an invoice manually (getting the blank page) so I guess the issue is outside of this PR?

**Comment by mariodian** [comment]
After an investigation I figured it is my mistake. Will fix it shortly.

**Comment by NicolasDorier** [comment]
Making new version with the fix. The previous version has not been released yet so that's fine.

**Comment by NicolasDorier** [comment]
Seems good! 

https://mainnet.demo.btcpayserver.org/apps/3utBTfSKkW4gK7aQMd2hW5Bh9Fpa/pos

Hey @mariodian I don't know how you are coding now, but [this guide](https://www.youtube.com/watch?v=ZePbMPSIvHM). Show you how to easily get btcpay running on localhost and debug/step by step etc...

**Comment by mariodian** [comment]
I’m doing it that way, the problem was that I tested it with the custom amount enabled so it always worked properly.

**Pull request merged**